src/idl/src/tree.c

```diff
@@ -983,6 +984,136 @@ static const char *describe_struct(const void *ptr)
   return "struct";
 }
 
+static int compare_id(const void *lhs, const void *rhs)
+{
+  if (*(const uint32_t*)lhs < *(const uint32_t*)rhs)
+    return -1;
+  else if (*(const uint32_t*)lhs > *(const uint32_t*)rhs)
+    return 1;
+  else
+    return 0;
+}
+
+idl_retcode_t
+propagate_autoid(
+  idl_pstate_t *pstate,
+  idl_node_t *list,
+  idl_autoid_t toset)
+{
+  assert(list);
+
+  idl_node_t *node = NULL;
+  idl_retcode_t ret = IDL_RETCODE_OK;
+  IDL_FOREACH(node, list) {
+    if (idl_is_module(node)
+     && !((idl_module_t*)node)->autoid.annotation) {
+      idl_module_t *mod = (idl_module_t*)node;
+      mod->autoid.value = toset;
+      ret = propagate_autoid(pstate, mod->definitions, toset);
+    }
+    if (idl_is_struct(node)
+     && !((idl_struct_t*)node)->autoid.annotation) {
+      idl_struct_t *str = (idl_struct_t*)node;
+      str->autoid.value = toset;
+      ret = propagate_autoid_struct(pstate, str);
+    }
+    if (ret)
+      break;
+  }
+
+  return ret;
+}
+
+idl_retcode_t
+propagate_autoid_struct(
+  idl_pstate_t *pstate,
+  idl_struct_t *str)
+{
+  assert(str);
+
+  size_t n_ids = 0;
+  uint32_t last = 0xFFFFFFFF;
+  bool last_is_inherited = false;
```

Comment:
I'd switch this to a `const idl_declarator_t *` and see if the ancestor of that matches the current.

Reply:
This approach has been changed.

Reply:
I switched to a different approach: not keeping track of the actual last id value, but just keeping the last declarator (which has the id value).